Switch to 'smbprotocol' library

malthe opened the pull request:
This change switches the Samba hook implementation to the [smbprotocol](https://pypi.org/project/smbprotocol/) library.

As discussed in #14054 the current implementation uses an unmaintained and outdated library.

In addition, the implementation suggested here conveniently exposes a large part of the user-friendly API provided by the underlying library. For example, to use the `smbclient.getxattr` function, one would call `samba_hook.getxattr` (with connection details being automatically provided).

---

closes: #14054

malthe committed:
Switch to 'smbprotocol' library

uranusjr commented on airflow/providers/samba/hooks/samba.py:
```diff
@@ -16,12 +16,43 @@
+    unlink,
+    utime,
+    walk,
+)
```
It’s probably better to use `import smbclient` and namespace everything under it. There are quite several function names that could be confusing an prone to coding errors (e.g. to those in `os`).

malthe replied:
I have updated the code now to reflect this. I think it's rare to see code that imports _from_ `os` and we do lose some import use tracking with this change, but as you say there are also some benefits.

uranusjr commented on airflow/providers/samba/hooks/samba.py:
```diff
@@ -16,12 +16,43 @@
 
 from airflow.hooks.base import BaseHook
 
+LOGGER = getLogger(__name__)
```
`BaseHook` is a `LoggingMixin` so we can just do `self.log.info(...)` instead of instantiating a new module-level logger. Although style around this is pretty diverse in the current code base so I’m not quite sure if `LoggingMixin` is preferred or not…

malthe replied:
I actually didn't know hooks were `LoggingMixin`. Nice!

uranusjr commented on airflow/providers/samba/hooks/samba.py:
```diff
@@ -31,52 +62,202 @@ class SambaHook(BaseHook):
+            "username": conn.login,
+            "password": conn.password,
+            "port": conn.port or 445,
+            "connection_cache": connection_cache,
```
Should this be `self._connection_cache`?

malthe replied:
Well, the variable is aliased locally to `connection_cache`.

uranusjr replied:
Oh OK, I misread.

malthe committed:
Use built-in logging mixin instance

malthe committed:
Import module instead and look up functions dynamically

This avoids contaminating the global namespace which could lead
to subtle errors.

uranusjr commented on airflow/providers/samba/hooks/samba.py:
```diff
@@ -16,67 +16,231 @@
+    @wraps(smbclient.link)
+    def link(self, src, dst, follow_symlinks=True):
+        return smbclient.link(
+            self._base_url + "/" + src,
```
Probably better to use `posixpath.join()` so users don’t need to deal with double slash issues.

It may be even better to change `_base_url` to something like

```python
def _format_url(self, endpoint):
    return f"//{posixpath.join(self._host, self._share, endpoint)}"
```

since the value of `_base_url` is never used along but always used to generate a full path (if I didn’t miss anything).

malthe replied:
Sounds good – I did exactly that.

malthe committed:
Use 'posixpath.join' to form UNC path

uranusjr reviewed (approved):
❤️

potiuk commented on airflow/providers/samba/hooks/samba.py:
```diff
@@ -16,67 +16,229 @@
+        }
+
+    def __enter__(self):
+        # This immediately connects to the host (which can be
```
If I read that correctly - it means that SambaHook MUST be used as ContextManager in some cases, otherwise there might some problems with initializing some parameters during constructor? This is not a usual pattern we have in Airflow for hooks (though I think it's nice pattern for Hooks) but I think some explanation is needed at least in the docstring explaining the difference between the two and when to use it? 

Also - would you mind to add a CHANGELOG.txt entry? Don't yet put the version (I will update it) but some backwards-compatibility notes are needed (how to migrate?)

malthe replied:
@potiuk where would those notes go? I don't see other changelog entries with such information. I have added a simple changelog entry for now and a note in the docstring about using it as a context manager.

potiuk replied:
You could see some of those in other providers - for example look at the Google provider: https://github.com/apache/airflow/blob/main/airflow/providers/google/CHANGELOG.rst - usually when there are braking changes that require more than one-line mentioning.

Even in Samba, we've added a warning when we removed "apply_default" (this was a global one and it actually did NOT really apply to samba but it went there as it was "all operator" change.

https://github.com/apache/airflow/blob/main/airflow/providers/samba/CHANGELOG.rst

malthe committed:
Add note that hook should be used as context manager and add changelog entry

malthe committed:
Fix compatibility with Python 3.7

See https://bugs.python.org/issue21269 for details.

kaxil commented on CHANGELOG.txt:
```diff
@@ -1,3 +1,10 @@
+""""""""""""
+
+- The Samba hook now uses the `smbprotocol` library, exposing a much
+  bigger set of functionality and supporting SMB2/3 protocols.
```
This is in wrong file -- it should be at https://github.com/apache/airflow/blob/main/airflow/providers/samba/CHANGELOG.rst

malthe replied:
@kaxil thanks – I have removed this entry again. I had actually already written a more detailed changelog entry in the provider changelog.

malthe committed:
Remove root changelog entry
